glottolog3/views.py

```diff
@@ -398,41 +399,144 @@ def bp_api_search(request):
 
 @view_config(
         route_name='glottolog.add_identifier',
-        request_method='POST',
         renderer='json')
 def add_identifier(request):
-    # TODO: Add validity checks for parameters and unit tests
-
-    gcode = request.json_body['glottocode']
-    lang = request.json_body['language']
-    name = request.json_body['name']
-    type = request.json_body['type']
-    desc = request.json_body['description']
+    gcode = request.matchdict['glottocode'].lower()
 
     languoid = DBSession.query(Language) \
                         .filter_by(id='{0}'.format(gcode)) \
                         .first()
 
-    identifier = Identifier(
-        (name, type, desc, lang),
-        id='{0}-{1}-{2}-{3}'.format(
-        slug(name), slug(type), slug(desc or ''), lang),
-        name=name,
-        type=type,
-        description=desc,
-        lang=lang)
-
+    if languoid == None:
+        request.response.status = 404 
+        return {'error': 'No language found with glottocode: {}'.format(gcode)}
+    
+    try:
+        identifier, errors = IdentifierSchema().load(request.json_body)
+    except (ValueError, ValidationError) as e:
+        request.response.status = 400
+        return {'error': '{}'.format(e)}
+    if errors:
+        request.response.status = 400
+        return {'error': errors}
+    
     try:
         DBSession.add(identifier)
         DBSession.add(
             LanguageIdentifier(language=languoid, identifier=identifier))
         DBSession.flush()
     except exc.SQLAlchemyError as e:
+        request.response.status = 400
         DBSession.rollback()
         return { 'error': '{}'.format(e) }
+    
+    # So we can use identifier object without session
+    DBSession.expunge(identifier)
+    transaction.commit()
+
+    return json.dumps(IdentifierSchema().dump(identifier))
+
+
+def query_identifier(type, name):
+    type = type.lower()
+    name = name.title() if type == 'name' else name.lower()
+
+    id_query = DBSession.query(Identifier) \
+                          .filter(and_(Identifier.name == name,
+                                       Identifier.type == type)) \
+
+    if id_query.count() == 0:
```

Comment:
Why not just return None if you're just going to put it in a try/catch every time?

Reply:
I thought it would be less redundant if I didn't put the entire message again in each function. I assumed that I'd have to check if it's None then print the entire message anyway.

Reply:
that's true. try/catches make me worried, though (I'm just nitpicking)